### PR TITLE
Alias builtin

### LINKED ISSUE #40: `alias` builtin

### PR DESCRIPTION
**Pull request by CraigNewlands**

### CODE REVIEW

**Comment by CraigNewlands** [comment]
works without quotes but will test with quotes and with spaces once we have sported the parser to handle the speech marks 
image

File: src/main.rs [hunk 1]
```diff
@@ -18,8 +22,64 @@ fn main() -> Result<(), std::io::Error> {
...
+            match builtins::echo(tokenized.args()) {
+                Ok(_) => {}
+                Err(e) => eprintln!("{}", e),
+            }
+        } else if tokenized.cmd() == "alias" {
+            match builtins::alias(tokenized.args(), &mut aliases) {
+                Ok(_) => {}
+                Err(e) => eprintln!("{}", e),
+            }
+        } else if let Some(alias_cmd) = aliases.get(tokenized.cmd()) {
```
Comment (siliconlad) [inline src/main.rs]: Wouldn't this be better off as a pre-processing step before we get to the switch statement?

**Comment by siliconlad** [comment]
This you might need to revise/force push to fix the merge conflicts too. Other than that (and my other comment) looks good!

File: src/main.rs (hunk 1 above)
Comment (CraigNewlands) [reply src/main.rs]: Yeah good point, save some time doing any of the switch statement

File: src/main.rs [hunk 2]
```diff
@@ -13,8 +16,29 @@ fn main() -> Result<(), std::io::Error> {
...
+                        eprintln!("Failed to execute command");
+                        continue;
+                    }
+                };
+
+            // Wait for the command to finish
+            match child.wait() {
+                Ok(_) => {}
+                Err(e) => eprintln!("{}", e),
+            }
```
Comment (siliconlad) [inline src/main.rs]: Ah, I think you misunderstood what I meant with my previous comment.

Because lines 25-40 are just running a plain old process, let's move lines 21-23 outside of the big if/else statement, but just make sure that if there are no matches for the alias, we don't modify the command at all.

You can imagine a sort of pipeline of modifications that we perform before we actually run it. One stage of that pipeline is inserting the relevant aliases.

**Review by siliconlad — COMMENTED** [review]
I rebased this branch so fix the diffs.

**Comment by siliconlad** [comment]
I've merged most of the other PRs so we should be able to revert the temp fix.

**Comment by CraigNewlands** [comment]
image

File: src/builtins.rs [hunk 3]
```diff
@@ -1,6 +1,13 @@
+use lazy_static::lazy_static;
+use std::collections::HashMap;
 use std::error::Error;
+use std::sync::Mutex;
 
-const BUILTINS: &[&str] = &["cd", "pwd", "exit", "echo"];
+lazy_static! {
```
Comment (CraigNewlands) [inline src/builtins.rs]: check if needed?

Comment (CraigNewlands) [reply src/builtins.rs]: removed as not needed

File: src/builtins.rs [hunk 4]
```diff
@@ -56,7 +61,44 @@ pub fn export(args: Vec<&str>) -> Result<String, Box<dyn Error>> {
...
+    unsafe {
+        if ALIASES.is_none() {
+            ALIASES = Some(HashMap::new());
+        }
+        ALIASES.as_mut().unwrap().insert(alias_name, alias_value);
+    }
+    Ok("".to_string())
+}
+
+pub fn expand_aliases(input: &str) -> String {
```
Comment (CraigNewlands) [inline src/builtins.rs]: Not sure if this should be in `main.rs` or not?

Comment (CraigNewlands) [reply src/builtins.rs]: As well as the `ALIASES` above? @siliconlad thoughts?

Comment (siliconlad) [inline src/builtins.rs]: This only detects aliases if they are the first word in a statement. Just realized that this isn't what we want I don't think because we also want aliases in sequences (and maybe pipelines?) to work as well I think:

```bash
alias_1 arg1; alias_2;
```

So this method should probably work after the tokenization step in `main.rs`.

File: src/builtins.rs [hunk 5]
```diff
@@ -1,6 +1,10 @@
+use std::collections::HashMap;
 use std::error::Error;
+use std::ptr::addr_of;
 
-const BUILTINS: &[&str] = &["cd", "pwd", "exit", "echo", "export"];
+static mut ALIASES: Option<HashMap<String, String>> = None;
```
Comment (siliconlad) [inline src/builtins.rs]: If we are doing alias substitution in `main.rs`, then it might be easier to just define it at the beginning there? Then you can pass it into `alias` and `expand_alias` and avoid all the unsafe code.

**Review by siliconlad — CHANGES_REQUESTED** [review]
Sorry, just realized that we probably want to support alias substitution in slightly more complicated scenarios haha :sweat_smile:

**Comment by CraigNewlands** [comment]
image

**Comment by siliconlad** [comment]
Should the "Hello world" have double quotes when greet is run?

**Comment by CraigNewlands** [comment]
> Should the "Hello world" have double quotes when greet is run?

It should not hahah, will have a look at that

**Comment by CraigNewlands** [comment]
image
After 2 months i think its finally done hahaha

File: src/main.rs [hunk 6]
```diff
@@ -70,3 +83,22 @@ fn main() -> Result<()> {
...
+            } else if let Some(cmd) = command.as_any_mut().downcast_mut::<ExternalCommand>() {
+                expand_aliases(&mut cmd.tokens, aliases);
+            }
+        }
+    } else if let Some(cmd) = tokenized.as_any_mut().downcast_mut::<BuiltinCommand>() {
+        expand_aliases(&mut cmd.tokens, aliases);
+    } else if let Some(cmd) = tokenized.as_any_mut().downcast_mut::<ExternalCommand>() {
+        expand_aliases(&mut cmd.tokens, aliases);
+    }
+}
```
Comment (siliconlad) [inline src/main.rs]: Do we not need a condition for pipelines too?

File: src/pipeline.rs [hunk 7]
```diff
@@ -44,6 +45,22 @@ impl Runnable for Pipeline {
...
+        self
+    }
+
+    fn to_string(&self) -> String {
+        self.commands
+            .iter()
+            .map(|cmd| cmd.to_string())
+            .collect::<Vec<String>>()
+            .join(" | ")
+    }
```
Comment (siliconlad) [inline src/pipeline.rs:63]: Is it worth implementing the [`Display`](https://doc.rust-lang.org/std/fmt/trait.Display.html) trait instead? Same for the others. Could be done in a separate issue maybe.

File: src/builtins.rs [hunk 8]
```diff
@@ -64,7 +72,44 @@ pub fn export(args: Vec<&str>) -> Result<String, Box<dyn Error>> {
...
+pub fn expand_aliases(tokens: &mut Vec<String>, aliases: &HashMap<String, String>) {
+    if let Some(first_token) = tokens.first() {
+        if let Some(alias_value) = aliases.get(first_token) {
+            let mut new_tokens: Vec<String> =
+                alias_value.split_whitespace().map(String::from).collect();
+            new_tokens.extend(tokens.drain(1..));
+            *tokens = new_tokens;
+        }
+    }
+}
```
Comment (siliconlad) [inline src/builtins.rs]: I feel like this should just return the new set of tokens. Since you're already creating a new `vec` of tokens `new_tokens`. Keep things a little more functional.

File: src/main.rs [hunk 9]
```diff
@@ -48,17 +57,21 @@ fn main() -> Result<()> {
...
-        debug!("Finished tokenizing...running command(s)");
+        debug!("Finished tokenizing...");
 
+        // Expand aliases
+        expand_aliases_in_tokenized(&mut tokenized, &aliases);
+
+        debug!("Running command(s)");
         // Run the command
-        match tokenized.run() {
+        match tokenized.run(&mut aliases) {
```
Comment (siliconlad) [inline src/main.rs]: Why do we do this? It doesn't look like this is used anywhere?

File: src/builtins.rs [hunk 10]
```diff
@@ -64,7 +72,44 @@ pub fn export(args: Vec<&str>) -> Result<String, Box<dyn Error>> {
...
+
+    aliases.insert(alias_name, tokenized_value);
+    Ok("".to_string())
+}
+
+pub fn expand_aliases(tokens: &mut Vec<String>, aliases: &HashMap<String, String>) {
+    if let Some(first_token) = tokens.first() {
+        if let Some(alias_value) = aliases.get(first_token) {
+            let mut new_tokens: Vec<String> =
+                alias_value.split_whitespace().map(String::from).collect();
```
Comment (siliconlad) [inline src/builtins.rs]: I wonder if there's a lossless way, because behaviour is slightly different (note the spaces):

```bash
> alias greet='echo " hello world "'
> greet
hello world
> echo " hello world "
 hello world 
```

Can't `HashMap` store `Vec<String>` for the value type?

**Review by siliconlad — CHANGES_REQUESTED** [review]
A couple of comments :)

File: src/command.rs [hunk 11]
```diff
@@ -70,7 +71,7 @@ pub fn runnable(tokens: Vec<Token>) -> Result<Box<dyn Runnable>, Box<dyn Error>>
 
 #[derive(Debug, Clone)]
 pub struct BuiltinCommand {
-    tokens: Vec<String>,
+    pub tokens: Vec<String>,
```
Comment (siliconlad) [inline src/command.rs]: Ideally we don't do this.

**Comment by CraigNewlands** [comment]
Completely changed my approach so now expanding the aliases before they are tokenized. This seems much simpler and means the string is still tokenized properly and we don't need to retokenize in the alias expansion if we did it after the tokenization. Seems to work for all the tests I've done - using piping and recursive aliases etc. What do you think?

**Comment by CraigNewlands** [comment]
Found a couple more things to fix

**Comment by CraigNewlands** [comment]
So many edge cases 😂

**Comment by siliconlad** [comment]
I think we should merge #82 so we can add all the edge-cases as tests.

**Comment by CraigNewlands** [comment]
> I think we should merge #82 so we can add all the edge-cases as tests.

Merged, will add some tests now for most of the edge cases I can think of.

**Comment by CraigNewlands** [comment]
@siliconlad can you think of a way I can get around this problem I'm having...

Currently if we do 
```
> alias greet='echo "Hello, World!"'
> greet
Hello, World!
> alias greet='echo "Hello, World!"' && alias
greet='echo "Hello, World!"'
> alias greet='echo "Hello, World!"' && greet
No such file or directory (os error 2)
> 
```

the last line doesn't work because `expand_aliases` is called before we `tokenize` and then add `alias greet='echo "Hello, World!"'` to the aliases hashmap, which is done in `tokenize.run`, so it is not expanded. Can you think of a good way of doing this?

**Comment by siliconlad** [comment]
What I might try (in `expand_aliases`) is the following sequence of steps:
1. Split string by `;`
2. Split string by `&&`
3. Split string by `|`
4. Replace the first word of each chunk by an alias if available

Would that work? Maybe you also need to deal with aliases using aliases. Perhaps that's just repeated application of the steps above until you detect no more aliases (might have to detect loops / set maximum alias depth).

**Comment by CraigNewlands** [comment]
It's not the prettiest code in the world but functionally I think it finally works. Going to try make it a bit nicer to read, especially the `expand_aliases` function lol

**Comment by CraigNewlands** [comment]
@siliconlad think we're good to go on this PR. Ready for review, hopefully for the final time lol

File: src/builtins.rs [hunk 12]
```diff
@@ -120,3 +125,34 @@ pub fn llm(args: Vec<String>, stdin: Option<ChildStdout>) -> Result<String, Box<
...
+            match aliases.get(&args[0]) {
+                Some(command) => {
+                    println!("{}='{}'", args[0], command);
+                    Ok("".to_string())
+                }
+                None => Err(format!("Alias '{}' not found", args[0]).into()),
+            }
+        }
+    } else {
+        Err("Usage: alias [name[=value] ...]".into())
```
Comment (siliconlad) [inline src/builtins.rs:156]: What's the `...`?

File: src/command.rs [hunk 13]
```diff
@@ -126,22 +133,45 @@ impl ShellCommand for BuiltinCommand {
...
                 dup2(pipe_out_w.as_raw_fd(), 1)?;
                 dup2(pipe_err_w.as_raw_fd(), 2)?;
-                self.run_builtin(stdin)?;
+                self.run_builtin(stdin, aliases)?;
+
+                // Send updated aliases to parent
+                let serialized_aliases = serde_json::to_string(aliases)?;
+                let mut alias_writer = unsafe { File::from_raw_fd(alias_w.into_raw_fd()) };
+                write!(alias_writer, "{}", serialized_aliases)?;
+
```
Comment (siliconlad) [inline src/command.rs:169]: This is really cool haha. I'm assuming that this is for this particular scenario:

```bash
> alias r='echo' | r Hello
Hello
```

But this doesn't work in Bash, so we don't necessarily need to support it? Might cut down on the complexity.

File: src/main.rs [hunk 14]
```diff
@@ -144,3 +154,66 @@ fn aishrc_path() -> Result<PathBuf, std::io::Error> {
...
+fn handle_alias_definition(temp_aliases: &mut HashMap<String, String>, args: &[String]) {
+    let full_command = args.join(" ");
+    if let Some(equals_pos) = full_command.find('=') {
+        let (alias_name, alias_value) = full_command.split_at(equals_pos);
+        temp_aliases.insert(
+            alias_name.to_string(),
+            alias_value[1..].trim_matches('\'').to_string(),
+        );
+    }
+}
```
Comment (siliconlad) [inline src/main.rs:205]: What is this for?

File: tests/integration.rs
Comment (siliconlad) [inline tests/integration.rs:1]: I guess it doesn't hurt to have all these tests, but probably worth having tests that execute aliases (you should be able to use `;`).

File: src/main.rs [hunk 15]
```diff
@@ -144,3 +154,66 @@ fn aishrc_path() -> Result<PathBuf, std::io::Error> {
...
+    let mut expansion_count = 0;
+
+    while let Some(expansion) = temp_aliases.get(&words[0]) {
+        let expanded_words: Vec<String> = expansion.split_whitespace().map(String::from).collect();
+        if expanded_words.is_empty() || expansion_count >= MAX_EXPANSIONS {
+            break;
+        }
+        words.splice(0..1, expanded_words);
+        expansion_count += 1;
+    }
```
Comment (siliconlad) [inline src/main.rs:218]: Reading the [reference manual](https://www.gnu.org/software/bash/manual/bash.html#Aliases), they actually don't do recursive expansions.

File: src/builtins.rs [hunk 16]
```diff
@@ -120,3 +125,34 @@ pub fn llm(args: Vec<String>, stdin: Option<ChildStdout>) -> Result<String, Box<
...
+        for (alias, command) in aliases {
+            println!("{}='{}'", alias, command);
+        }
+        Ok("".to_string())
+    } else if args.len() == 1 {
+        let parts: Vec<&str> = args[0].splitn(2, '=').collect();
+        if parts.len() == 2 {
+            let alias = parts[0];
+            let command = parts[1];
+            aliases.insert(alias.to_string(), command.to_string());
```
Comment (siliconlad) [inline src/builtins.rs:144]: The [reference manual](https://www.gnu.org/software/bash/manual/bash.html#Aliases) does not allow special characters in the name. Not sure if we want to enforce that or not :)

**Review by siliconlad — COMMENTED** [review]
I guess the main thing is it works.

I left some comments mainly about what bash does and whether we want to behave the same way or not. I think some of it will allow us to cut down on the complexity that `alias` adds (e.g. in `pipe`). What do you think?

File: src/main.rs (hunk 14 above)
Comment (CraigNewlands) [reply src/main.rs:205]: It's solving the problem where we have something like `alias greet='echo \"Hello, World!\"' && greet` because expand tokens gets called before we add greet to the aliases, the second greet after the && doesn't get expanded so this checks if we have an alias on the left of the command and if we do then it expands the right hand side. So basically all handle_alias_definition is doing is getting the alias greet and the expansion of this which is echo \"Hello, World!\", this is then returned and used in expand_command_aliases to turn the whole command into `alias greet='echo \"Hello, World!\"' && alias greet='echo \"Hello, World!\"`

Comment (CraigNewlands) [reply src/main.rs:205]: This could probably be made into one function as we are doing the same thing in the alias buitlin command.
